app/gui/src/controller/graph/widget/configuration.rs

```diff
@@ -51,34 +52,88 @@ pub fn deserialize_widget_definitions(
 
 /// Convert a widget definition from the engine response into a IDE internal widget configuration
 /// struct. See [`widget::Configuration`] for more information.
-fn to_configuration(resp: response::WidgetDefinition) -> widget::Configuration {
-    widget::Configuration { display: resp.display, kind: to_kind(resp.inner), has_port: true }
+fn to_configuration(
+    resp: response::WidgetDefinition,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> widget::Configuration {
+    widget::Configuration {
+        display:  resp.display,
+        kind:     to_kind(resp.inner, db, parser),
+        has_port: true,
+    }
 }
 
-fn to_kind(inner: response::WidgetKindDefinition) -> widget::DynConfig {
+fn to_kind(
+    inner: response::WidgetKindDefinition,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> widget::DynConfig {
     match inner {
-        response::WidgetKindDefinition::SingleChoice { label, values } =>
+        response::WidgetKindDefinition::SingleChoice { label, values } => {
+            let (entries, arguments) = to_entries_and_arguments(values, db, parser);
             widget::single_choice::Config {
-                label:   label.map(Into::into),
-                entries: Rc::new(to_entries(values)),
+                label: label.map(Into::into),
+                entries: Rc::new(entries),
+                arguments,
             }
-            .into(),
+            .into()
+        }
         response::WidgetKindDefinition::ListEditor { item_widget, item_default } =>
             widget::list_editor::Config {
-                item_widget:  Some(Rc::new(to_configuration(*item_widget))),
+                item_widget:  Some(Rc::new(to_configuration(*item_widget, db, parser))),
                 item_default: ImString::from(item_default).into(),
             }
             .into(),
         _ => widget::label::Config::default().into(),
     }
 }
 
-fn to_entries(choices: Vec<response::Choice>) -> Vec<widget::Entry> {
-    choices.into_iter().map(to_entry).collect()
+fn to_entries_and_arguments(
+    choices: Vec<response::Choice>,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> (Vec<widget::Entry>, Vec<(usize, ImString, widget::Configuration)>) {
+    let mut args = Vec::new();
+
+    let expressions = choices.iter().map(|c| c.value.as_ref());
+    let as_tags = argument_tag_values(expressions, db, parser);
+
+    let entries = choices
+        .into_iter()
+        .enumerate()
+        .zip(as_tags)
+        .map(|((i, c), tag)| to_entry(c, i, db, parser, tag, &mut args))
```

Comment:
`i` and `c` are rather bad names here as it's really hard to understand from the context code what they refer to.